aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,288 @@
+  // std::cout << "MaxUnpooling2d_forward_cpu called" << "\n";
+  AT_CHECK(
+      self.ndimension() == 4,
+      "Input to MaxUnpooling2d should be a NCHW Tensor",
+      self.sizes());
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (height, width) in output_size");
+  auto output = at::zeros(
+      {self.size(0), self.size(1), output_size[0], output_size[1]},
+      self.options());
+  MaxUnpooling2d_forward_out_cpu(output, self, indices, output_size);
+  return output;
+};
+
+Tensor MaxUnpooling2d_backward_cpu(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_ERROR("not implemented");
+}
+
+// stopgap until GPU version is implemented
+at::Tensor& MaxUnpooling2d_forward_out_cuda(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  // std::cout << "MaxUnpooling2d_forward_out_gpu called" << "\n";
+  return at::_thnn_max_unpool2d_out(output, self, indices, output_size);
+}
+
+// stopgap until GPU version is implemented
+at::Tensor MaxUnpooling2d_forward_cuda(
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  return at::_thnn_max_unpool2d(self, indices, output_size);
+}
+
+template <typename scalar_t>
+at::Tensor MaxUnpooling3d_forward_out_cpu_(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& indices,
+    int64_t oT,
+    int64_t oW,
+    int64_t oH,
+    int64_t dT,
+    int64_t dW,
+    int64_t dH,
+    int64_t pT,
+    int64_t pW,
+    int64_t pH) {
+  auto nBatch = input.size(0);
+  auto nSlices = input.size(1);
+
+  auto dimw = 4;
+  auto dimh = 3;
+  auto dimt = 2;
+
+  auto iT = input.size(dimt);
+  auto iH = input.size(dimh);
+  auto iW = input.size(dimw);
+
+  // TODO: zero out output inplace
+  auto* input_data = input.data<scalar_t>();
+  auto* output_data = output.data<scalar_t>();
+  auto* indices_data = indices.data<int64_t>();
+
+  for (auto p = 0; p < nBatch; p++) {
+    auto inputOffset = p * nSlices * iT * iW * iH;
+    auto outputOffset = p * nSlices * oT * oW * oH;
+
+    for (auto k = 0; k < nSlices; k++) {
+      auto finalInputOffset = inputOffset + k * iT * iW * iH;
+      auto finalOutputOffset = outputOffset + k * oT * oW * oH;
+
+      auto* output_p_k = output_data + finalOutputOffset;
+      auto* input_p_k = input_data + finalInputOffset;
+      auto* ind_p_k = indices_data + finalInputOffset;
+      int maxp;
+      for (auto t = 0; t < iT; t++) {
+        for (auto i = 0; i < iH; i++) {
+          for (auto j = 0; j < iW; j++) {
+            auto index = t * iH * iW + i * iW + j;
+            maxp = ind_p_k[index];
+            if (maxp < 0 || maxp >= oT * oW * oH) {
+              AT_ERROR("Invalid index");
+            } else {
+              output_p_k[maxp] = input_p_k[index];
+            }
+          }
+        }
+      }
+    }
+  }
+  return output;
+}
+
+at::Tensor& MaxUnpooling3d_forward_out_cpu(
```

Comment:
The Python that calls this function always calculates output_size from the other variables that are passed in (https://github.com/pytorch/pytorch/blob/a66669a110a6eeeef8e08293174138ca87dc1a66/torch/nn/functional.py#L550), which means that `stride` and `padding` are redundant here and never get used - at least that's what I'm seeing here.